File: core/src/main/java/com/egoal/darkestpixeldungeon/actors/blobs/WaterOfTransmutation.java
/*
 * Pixel Dungeon
 * Copyright (C) 2012-2015  Oleg Dolya
 *
 * Shattered Pixel Dungeon
 * Copyright (C) 2014-2016 Evan Debenham
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 */
package com.egoal.darkestpixeldungeon.actors.blobs;

import com.egoal.darkestpixeldungeon.Journal;
import com.egoal.darkestpixeldungeon.DarkestPixelDungeon;
import com.egoal.darkestpixeldungeon.effects.BlobEmitter;
import com.egoal.darkestpixeldungeon.effects.Speck;
import com.egoal.darkestpixeldungeon.items.Generator;
import com.egoal.darkestpixeldungeon.items.artifacts.Artifact;
import com.egoal.darkestpixeldungeon.items.potions.PotionOfStrength;
import com.egoal.darkestpixeldungeon.items.scrolls.Scroll;
import com.egoal.darkestpixeldungeon.items.scrolls.ScrollOfMagicalInfusion;
import com.egoal.darkestpixeldungeon.items.weapon.Weapon;
import com.egoal.darkestpixeldungeon.items.weapon.melee.MagesStaff;
import com.egoal.darkestpixeldungeon.messages.Messages;
import com.egoal.darkestpixeldungeon.plants.Plant;
import com.egoal.darkestpixeldungeon.items.Item;
import com.egoal.darkestpixeldungeon.items.potions.Potion;
import com.egoal.darkestpixeldungeon.items.potions.PotionOfMight;
import com.egoal.darkestpixeldungeon.items.rings.Ring;
import com.egoal.darkestpixeldungeon.items.scrolls.ScrollOfUpgrade;
import com.egoal.darkestpixeldungeon.items.wands.Wand;
import com.egoal.darkestpixeldungeon.items.weapon.melee.MeleeWeapon;
import com.watabou.utils.Random;

public class WaterOfTransmutation extends WellWater {

  @Override
  protected Item affectItem(Item item) {

    if (item instanceof MagesStaff) {
      item = changeStaff((MagesStaff) item);
    } else if (item instanceof MeleeWeapon) {
      item = changeWeapon((MeleeWeapon) item);
    } else if (item instanceof Scroll) {
      item = changeScroll((Scroll) item);
    } else if (item instanceof Potion) {
      item = changePotion((Potion) item);
    } else if (item instanceof Ring) {
      item = changeRing((Ring) item);
    } else if (item instanceof Wand) {
      item = changeWand((Wand) item);
    } else if (item instanceof Plant.Seed) {
      item = changeSeed((Plant.Seed) item);
    } else if (item instanceof Artifact) {
      item = changeArtifact((Artifact) item);
    } else {
      item = null;
    }

    if (item != null) {
      Journal.remove(Journal.Feature.WELL_OF_TRANSMUTATION);
    }

    return item;

  }

  @Override
  public void use(BlobEmitter emitter) {
    super.use(emitter);
    emitter.start(Speck.factory(Speck.CHANGE), 0.2f, 0);
  }

  private MagesStaff changeStaff(MagesStaff staff) {
    Class<? extends Wand> wandClass = staff.wandClass();

    if (wandClass == null) {
      return null;
    } else {
      Wand n;
      do {
        n = (Wand) Generator.random(Generator.Category.WAND);
      } while (n.getClass() == wandClass);
      n.level(0);
      staff.imbueWand(n, null);
    }

    return staff;
  }

  private Weapon changeWeapon(MeleeWeapon w) {

    Weapon n;
    Generator.Category c = Generator.wepTiers[w.tier - 1];

    do {
      try {
        n = (Weapon) c.classes[Random.chances(c.probs)].newInstance();
      } catch (Exception e) {
        DarkestPixelDungeon.reportException(e);
        return null;
      }
    } while (!(n instanceof MeleeWeapon) || n.getClass() == w.getClass());

    int level = w.level();
    if (level > 0) {
      n.upgrade(level);
    } else if (level < 0) {
      n.degrade(-level);
    }

    n.enchantment = w.enchantment;
    n.levelKnown = w.levelKnown;
    n.cursedKnown = w.cursedKnown;
    n.cursed = w.cursed;
    n.imbue = w.imbue;

    return n;

  }

  private Ring changeRing(Ring r) {
    Ring n;
    do {
      n = (Ring) Generator.random(Generator.Category.RING);
    } while (n.getClass() == r.getClass());

    n.level(0);

    int level = r.level();
    if (level > 0) {
      n.upgrade(level);
    } else if (level < 0) {
      n.degrade(-level);
    }

    n.levelKnown = r.levelKnown;
    n.cursedKnown = r.cursedKnown;
    n.cursed = r.cursed;

    return n;
  }

  private Artifact changeArtifact(Artifact a) {
    Artifact n = Generator.randomArtifact();

    if (n != null) {
      n.cursedKnown = a.cursedKnown;
      n.cursed = a.cursed;
      n.levelKnown = a.levelKnown;
      n.transferUpgrade(a.visiblyUpgraded());
    }

    return n;
  }

  private Wand changeWand(Wand w) {

    Wand n;
    do {
      n = (Wand) Generator.random(Generator.Category.WAND);
    } while (n.getClass() == w.getClass());

    n.level(0);
    n.upgrade(w.level());

    n.levelKnown = w.levelKnown;
    n.cursedKnown = w.cursedKnown;
    n.cursed = w.cursed;

    return n;
  }

  private Plant.Seed changeSeed(Plant.Seed s) {

    Plant.Seed n;

    do {
      n = (Plant.Seed) Generator.random(Generator.Category.SEED);
    } while (n.getClass() == s.getClass());

    return n;
  }

  private Scroll changeScroll(Scroll s) {
    if (s instanceof ScrollOfUpgrade) {

      return new ScrollOfMagicalInfusion();

    } else if (s instanceof ScrollOfMagicalInfusion) {

      return new ScrollOfUpgrade();

    } else {

      Scroll n;
      do {
        n = (Scroll) Generator.random(Generator.Category.SCROLL);
      } while (n.getClass() == s.getClass());
      return n;
    }
  }

  private Potion changePotion(Potion p) {
    if (p instanceof PotionOfStrength) {

      return new PotionOfMight();

    } else if (p instanceof PotionOfMight) {

      return new PotionOfStrength();

    } else {

      Potion n;
      do {
        n = (Potion) Generator.random(Generator.Category.POTION);
      } while (n.getClass() == p.getClass());
      return n;
    }
  }

  @Override
  public String tileDesc() {
    return Messages.get(this, "desc");
  }
}

File: core/src/main/java/com/egoal/darkestpixeldungeon/actors/blobs/WaterOfTransmutation.kt
package com.egoal.darkestpixeldungeon.actors.blobs

import com.egoal.darkestpixeldungeon.Journal
import com.egoal.darkestpixeldungeon.effects.BlobEmitter
import com.egoal.darkestpixeldungeon.effects.Speck
import com.egoal.darkestpixeldungeon.items.Item
import com.egoal.darkestpixeldungeon.items.KGenerator
import com.egoal.darkestpixeldungeon.items.artifacts.Artifact
import com.egoal.darkestpixeldungeon.items.helmets.Helmet
import com.egoal.darkestpixeldungeon.items.potions.Potion
import com.egoal.darkestpixeldungeon.items.potions.PotionOfMight
import com.egoal.darkestpixeldungeon.items.potions.PotionOfStrength
import com.egoal.darkestpixeldungeon.items.rings.Ring
import com.egoal.darkestpixeldungeon.items.scrolls.Scroll
import com.egoal.darkestpixeldungeon.items.scrolls.ScrollOfUpgrade
import com.egoal.darkestpixeldungeon.items.wands.Wand
import com.egoal.darkestpixeldungeon.items.weapon.Weapon
import com.egoal.darkestpixeldungeon.items.weapon.melee.MagesStaff
import com.egoal.darkestpixeldungeon.items.weapon.melee.MeleeWeapon
import com.egoal.darkestpixeldungeon.messages.Messages
import com.egoal.darkestpixeldungeon.plants.Plant

class WaterOfTransmutation : WellWater() {
    override fun affectItem(item: Item): Item? {
        val new = when (item) {
            is MagesStaff -> changeStaff(item)
            is MeleeWeapon -> changeWeapon(item)
            is Scroll -> changeScroll(item)
            is Potion -> changePotion(item)
            is Ring -> changeRing(item)
            is Wand -> changeWand(item)
            is Plant.Seed -> changeSeed(item)
            is Artifact -> changeArtifact(item)
            is Helmet -> changeHelmet(item)
            else -> null
        }

        if (new != null) Journal.remove(Journal.Feature.WELL_OF_TRANSMUTATION)
        return new
    }

    private fun changeStaff(staff: MagesStaff): MagesStaff? {
        return staff.wandClass()?.let {
            var wand = KGenerator.WAND.generate() as Wand
            while (wand.javaClass == it)
                wand = KGenerator.WAND.generate() as Wand
            wand.level(0)
            staff.imbueWand(wand, null)

            staff
        }
    }

    private fun changeWeapon(weap: MeleeWeapon): Weapon? {
        var newWeap = KGenerator.WEAPON.tier(weap.tier).generate()
        while (newWeap !is MeleeWeapon || weap.javaClass == newWeap.javaClass)
            newWeap = KGenerator.WEAPON.tier(weap.tier).generate()

        val n = newWeap as Weapon
        n.level(0)
        val lvl = weap.level()
        if (lvl > 0) n.upgrade(lvl) else if (lvl < 0) n.degrade(-lvl)

        n.enchantment = weap.enchantment
        n.levelKnown = weap.levelKnown
        n.cursedKnown = weap.cursedKnown
        n.cursed = weap.cursed
        n.imbue = weap.imbue

        return n
    }

    private fun changeRing(r: Ring): Ring? {
        var n = KGenerator.RING.generate() as Ring
        while (n.javaClass == r.javaClass) n = KGenerator.RING.generate() as Ring
        n.level(0)
        
        val lvl = r.level()
        if (lvl > 0) n.upgrade(lvl) else if (lvl < 0) n.degrade(-lvl)

        n.levelKnown = r.levelKnown
        n.cursedKnown = r.cursedKnown
        n.cursed = r.cursed

        return n
    }

    private fun changeArtifact(a: Artifact): Artifact? {
        val n = KGenerator.ARTIFACT.generate()

        if (n !is Artifact) return null // if we run out of artifacts, do nothing

        n.cursedKnown = a.cursedKnown
        n.cursed = a.cursed
        n.levelKnown = a.levelKnown
        n.transferUpgrade(a.visiblyUpgraded())

        return n
    }

    private fun changeWand(w: Wand): Wand? {
        var n = KGenerator.WAND.generate() as Wand
        while (n.javaClass == w.javaClass) n = KGenerator.WAND.generate() as Wand
        n.level(0)
        
        n.upgrade(w.level())

        n.levelKnown = w.levelKnown
        n.cursedKnown = w.cursedKnown
        n.cursed = w.cursed

        return n
    }

    private fun changeSeed(s: Plant.Seed): Plant.Seed? {
        var n = KGenerator.SEED.generate() as Plant.Seed
        while (n.javaClass == s.javaClass) n = KGenerator.SEED.generate() as Plant.Seed

        return n
    }

    private fun changeScroll(s: Scroll): Scroll? = if (s is ScrollOfUpgrade) null else {
        var n = KGenerator.SCROLL.generate() as Scroll
        while (n.javaClass == s.javaClass) n = KGenerator.SCROLL.generate() as Scroll
        n
    }

    private fun changePotion(p: Potion): Potion? = when (p) {
        is PotionOfStrength -> PotionOfMight()
        is PotionOfMight -> PotionOfStrength()
        else -> {
            var n = KGenerator.POTION.generate() as Potion
            while (n.javaClass == p.javaClass) n = KGenerator.POTION.generate() as Potion
            n
        }
    }

    private fun changeHelmet(h: Helmet): Helmet? {
        var n = KGenerator.HELMET.generate() as Helmet
        while (n.javaClass == h.javaClass) n = KGenerator.HELMET.generate() as Helmet
        return n
    }

    override fun use(emitter: BlobEmitter) {
        super.use(emitter)
        emitter.pour(Speck.factory(Speck.CHANGE), 0.3f)
    }
    
    override fun tileDesc(): String = Messages.get(this, "desc")
}